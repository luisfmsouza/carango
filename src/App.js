import React, { useState } from "react";
import axios from "axios";
import set from "lodash/fp/set";
import isAlpha from "validator/lib/isAlpha";
import isEmail from "validator/lib/isEmail";
import isMobilePhone from "validator/lib/isMobilePhone";
import isNumeric from "validator/lib/isNumeric";

import PromotionScreen from "./PromotionScreen";
import isValidVoucherCode from "./promotion-codes/validate";
import { hasFormError, hasFormData } from "./selector";

const initialState = {
  firstName: { value: "", hasError: false, type: "string" },
  surname: { value: "", hasError: false, type: "string" },
  email: { value: "", hasError: false, type: "email" },
  phoneNumber: { value: "", hasError: false, type: "phone" },
  voucherCode: { value: "", hasError: false, type: "code" },
  acceptPrivacy: { value: false, hasError: false, type: "checkbox" }
};

export default WrappedComponent => {
  const [formState, setFormState] = useState(initialState);
  const [screen, setScreen] = useState("");

  const handleInputChange = event => {
    const input = event.target;
    const { type } = formState[input.id];
    const value = type === "checkbox" ? input.checked : input.value;
    const newState = set(`${input.id}.value`, value, formState);

    setFormState(newState);
  };

  const handleInputBlur = event => {
    const input = event.target;
    const { type, value } = formState[input.id];
    let hasError = false;

    if (type === "string") {
      hasError = !isAlpha(value) || value.length < 2;
    } else if (type === "email") {
      hasError = !isEmail(value);
    } else if (type === "phone") {
      hasError = !isMobilePhone(value, ["nl-NL", "pt-BR"]);
    } else if (type === "code") {
      hasError = !isNumeric(value) || !isValidVoucherCode(value);
    }

    const newState = set(`${input.id}.hasError`, hasError, formState);

    setFormState(newState);
  };

  const handleSubmit = event => {
    event.preventDefault();

    if (!hasFormError(formState) && hasFormData(formState)) {
      const instance = axios.create({
        baseURL: "https://some-domain.com/api/",
        timeout: 1000,
        headers: { "Content-Type": "application/json" }
      });

      instance
        .post("http://localhost:3004/form", {
          firstName: formState.firstName.value,
          surname: formState.surname.value,
          email: formState.email.value,
          phoneNumber: formState.phoneNumber.value,
          voucherCode: formState.voucherCode.value,
          acceptPrivacy: formState.acceptPrivacy.value
        })
        .then(function(response) {
          // Returns `true` ("You win") once every 10 executions
          if (Math.trunc(Math.random() * 10) === 1) {
            setScreen("win");
          } else {
            setScreen("lose");
          }
        })
        .catch(function(error) {
          setScreen("lose");
        });
    }
  };

  const shouldBlockFormSubmission =
    hasFormError(formState) || !hasFormData(formState);

  return (
    <PromotionScreen
      screen={screen}
      formState={formState}
      shouldBlockFormSubmission={shouldBlockFormSubmission}
      handleInputChange={handleInputChange}
      handleInputBlur={handleInputBlur}
      handleSubmit={handleSubmit}
    />
  );
};
