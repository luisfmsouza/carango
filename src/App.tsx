import React, { FC } from "react";

import logo from "./assets/images/car-next-logo.png";
import { ReactComponent as IconRight } from "./assets/icons/arrow-right.svg";

export const App: FC = () => (
  <>
    <header>
      <a href="#">
        <img src={logo} alt="CarNext logo" title="CarNext logo" />
      </a>
    </header>
    <section>
      <h1>Get a free year of driving!</h1>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. In erat eros,
        finibus ac ullamcorper sit amet, tempus quis lacus. Quisque eleifend
        semper pretium. Sed rutrum odio lectus, id maximus nunc rhoncus sed.
        Praesent quis suscipit dui. Lorem ipsum dolor sit amet, consectetur
        adipiscing elit. Integer orci dui, rutrum eu feugiat lobortis, tristique
        sollicitudin est. Nulla tincidunt semper nibh at pulvinar. Etiam tempor
        quam vitae neque ornare efficitur. Morbi id egestas est, at viverra
        orci. Fusce eleifend placerat ligula, eget viverra lectus facilisis
        vitae. Praesent placerat tortor a mi laoreet, quis lobortis nunc
        eleifend.
      </p>
    </section>
    <section>
      <h2>Get a car leasing offer</h2>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. In erat eros,
        finibus ac ullamcorper sit amet, tempus quis lacus.
      </p>
      <form>
        <label htmlFor="form--first-name">First name</label>
        <input
          id="form--first-name"
          placeholder="Enter first name"
          type="text"
        />
        <label htmlFor="form--surname">Surname</label>
        <input id="form--surname" placeholder="Enter surname" type="text" />
        <label htmlFor="form--email">Email address</label>
        <input
          id="form--email"
          placeholder="Enter your email address"
          type="email"
        />
        <label htmlFor="form--phone">Phone number</label>
        <input
          id="form--phone"
          placeholder="Enter your phone number"
          type="tel"
        />
        <label htmlFor="form--voucher">Voucher code</label>
        <input
          id="form--voucher"
          placeholder="Enter your voucher code"
          type="number"
        />
        <input id="form--policy" type="checkbox" />
        <label htmlFor="form--policy">Accept Privacy Settlement</label>
        <button type="submit">
          Get Offer
          <IconRight />
        </button>
      </form>
    </section>
  </>
);
