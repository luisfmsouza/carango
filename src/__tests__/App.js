import React from "react";
import renderer from "react-test-renderer";

import App from "../App";

describe("App", () => {
  it("renders the App component", () => {
    const actual = renderer.create(<App />).toJSON();

    expect(actual).toMatchSnapshot();
  });
});
