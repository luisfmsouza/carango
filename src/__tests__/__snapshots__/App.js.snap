// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the App component 1`] = `
<div
  className="sc-VigVT cVEKMV"
>
  <header
    className="sc-bdVaJa blMrvh"
  >
    <a
      href="#"
    >
      <img
        alt="CarNext logo"
        className="sc-bwzfXH hqJQNA"
        src="car-next-logo.png"
        title="CarNext logo"
      />
    </a>
  </header>
  <div
    className="sc-jTzLTM JpSzt"
  >
    <section
      className="sc-fjdhpX nsimD"
    >
      <h1
        className="sc-jzJRlG iUwjYv"
      >
        Get a free year of driving!
      </h1>
      <p
        className="sc-cSHVUG dmNwFQ"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. In erat eros, finibus ac ullamcorper sit amet, tempus quis lacus. Quisque eleifend semper pretium. Sed rutrum odio lectus, id maximus nunc rhoncus sed. Praesent quis suscipit dui. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer orci dui, rutrum eu feugiat lobortis, tristique sollicitudin est. Nulla tincidunt semper nibh at pulvinar. Etiam tempor quam vitae neque ornare efficitur. Morbi id egestas est, at viverra orci. Fusce eleifend placerat ligula, eget viverra lectus facilisis vitae. Praesent placerat tortor a mi laoreet, quis lobortis nunc eleifend.
      </p>
    </section>
    <section
      className="sc-fjdhpX nsimD"
    >
      <h1
        className="sc-jzJRlG iUwjYv"
      >
        Get a car leasing offer
      </h1>
      <p
        className="sc-cSHVUG dmNwFQ"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. In erat eros, finibus ac ullamcorper sit amet, tempus quis lacus.
      </p>
      <form
        className="sc-htpNat KtFsv"
      >
        <div
          className="sc-bxivhb cSLsnY"
        >
          <div
            className="sc-ifAKCX jpecgC"
          >
            <label
              className="sc-EHOje bPSVmV"
              htmlFor="form--first-name"
            >
              First name
            </label>
            <input
              className="sc-bZQynM kbmjkK"
              id="form--first-name"
              placeholder="Enter first name"
              type="text"
            />
          </div>
          <div
            className="sc-ifAKCX jpecgC"
          >
            <label
              className="sc-EHOje bPSVmV"
              htmlFor="form--surname"
            >
              Surname
            </label>
            <input
              className="sc-bZQynM kbmjkK"
              id="form--surname"
              placeholder="Enter surname"
              type="text"
            />
          </div>
        </div>
        <div
          className="sc-ifAKCX jpecgC"
        >
          <label
            className="sc-EHOje bPSVmV"
            htmlFor="form--email"
          >
            Email address
          </label>
          <input
            className="sc-bZQynM kbmjkK"
            id="form--email"
            placeholder="Enter your email address"
            type="email"
          />
        </div>
        <div
          className="sc-ifAKCX jpecgC"
        >
          <label
            className="sc-EHOje bPSVmV"
            htmlFor="form--phone"
          >
            Phone number
          </label>
          <input
            className="sc-bZQynM kbmjkK"
            id="form--phone"
            placeholder="Enter your phone number"
            type="tel"
          />
        </div>
        <div
          className="sc-ifAKCX jpecgC"
        >
          <label
            className="sc-EHOje bPSVmV"
            htmlFor="form--voucher"
          >
            Voucher code
          </label>
          <input
            className="sc-bZQynM kbmjkK"
            id="form--voucher"
            placeholder="Enter your voucher code"
            type="number"
          />
        </div>
        <div
          className="sc-ifAKCX jpecgC"
        >
          <input
            className="sc-htoDjs jtEFYT"
            id="form--policy"
            type="checkbox"
          />
          <span
            className="sc-dnqmqq ekAjtq"
          />
          <label
            className="sc-gzVnrw hNTdaM"
            htmlFor="form--policy"
          >
            Accept Privacy Settlement
          </label>
        </div>
        <button
          className="sc-iwsKbI dWNEbR"
          type="submit"
        >
          <span
            className="sc-gZMcBi dKOCRF"
          >
            Get Offer
          </span>
          <svg
            className="sc-gqjmRU fPOneR"
          >
            arrow-right.svg
          </svg>
        </button>
      </form>
    </section>
  </div>
</div>
`;
